server/src/test/java/org/elasticsearch/gateway/MetaDataStateFormatTests.java

```diff
@@ -344,8 +458,22 @@ public DummyState fromXContent(XContentParser parser) throws IOException {
         }
 
         @Override
-        protected Directory newDirectory(Path dir) throws IOException {
-            MockDirectoryWrapper  mock = new MockDirectoryWrapper(random(), super.newDirectory(dir));
+        protected Directory newDirectory(Path dir) {
+            MockDirectoryWrapper mock = newMockFSDirectory(dir);
+            if (injectFailure.get()) {
+                String failMethod = randomFrom(failureMethods);
+                MockDirectoryWrapper.Failure fail = new MockDirectoryWrapper.Failure() {
+                    @Override
+                    public void eval(MockDirectoryWrapper dir) throws IOException {
+                        for (StackTraceElement e : Thread.currentThread().getStackTrace()) {
+                                if (failMethod.equals(e.getMethodName())) {
+                                    throw new MockDirectoryWrapper.FakeIOException();
```

Comment:
Yes, it will fail on the first method call which name equals to one of randomly selected from failure methods.
For example, if you look at "testFailWriteAndReadPreviousState" you will find that failure methods are the following: "createOutput", "sync", "rename". If we run the sufficient number of iterations of the test, we can be sure that we test failures at all possible stages.